File: web-admin/src/api/conversion.ts
import { api } from './client';

export interface AssetConversion {
    id: string;
    request_number: string;
    asset_id: string;
    title: string;
    status: 'pending' | 'approved' | 'rejected' | 'executed' | 'cancelled';
    from_category_id: string;
    to_category_id: string;
    target_specifications?: any;
    conversion_cost: number;
    cost_treatment: 'capitalize' | 'expense';
    reason: string;
    notes?: string;
    requested_by: string;
    approved_by?: string;
    executed_by?: string;
    request_date: string;
    approval_date?: string;
    execution_date?: string;
    created_at?: string;
    updated_at?: string;
}

export interface CreateConversionRequest {
    asset_id: string;
    title: string;
    to_category_id: string;
    target_specifications?: any;
    conversion_cost: number;
    cost_treatment: 'capitalize' | 'expense';
    reason: string;
}

export interface ExecuteConversionRequest {
    notes?: string;
}

export const conversionApi = {
    createRequest: async (assetId: string, data: CreateConversionRequest) => {
        const response = await api.post<any>(`/assets/${assetId}/conversion-requests`, data);
        return response.data;
    },

    getPendingRequests: async () => {
        const response = await api.get<any>('/conversion-requests/pending');
        return response.data?.data || [];
    },

    getAssetConversions: async (assetId: string) => {
        const response = await api.get<any>(`/assets/${assetId}/conversion-requests`);
        return response.data?.data || [];
    },

    getById: async (id: string) => {
        const response = await api.get<any>(`/conversion-requests/${id}`);
        return response.data?.data;
    },

    approveRequest: async (id: string) => {
        const response = await api.put<any>(`/conversion-requests/${id}/approve`);
        return response.data;
    },

    rejectRequest: async (id: string) => {
        const response = await api.put<any>(`/conversion-requests/${id}/reject`);
        return response.data;
    },

    executeConversion: async (id: string, data: ExecuteConversionRequest) => {
        const response = await api.post<any>(`/conversion-requests/${id}/execute`, data);
        return response.data;
    },
};

File: src/domain/entities/conversion.rs
//! Asset Conversion Entity

use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct AssetConversion {
    pub id: Uuid,
    pub request_number: String,
    pub asset_id: Uuid,
    pub title: String,

    pub status: String, // pending, approved, rejected, executed, cancelled

    pub from_category_id: Option<Uuid>, // Optional because created from join or logic
    pub to_category_id: Uuid,
    pub target_specifications: Option<JsonValue>,

    pub conversion_cost: Decimal,
    pub cost_treatment: String, // capitalize, expense

    pub reason: Option<String>,
    pub notes: Option<String>,

    pub requested_by: Uuid,
    pub approved_by: Option<Uuid>,
    pub executed_by: Option<Uuid>,

    pub request_date: Option<DateTime<Utc>>,
    pub approval_date: Option<DateTime<Utc>>,
    pub execution_date: Option<DateTime<Utc>>,

    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
